### PR TITLE
[WIP] Visible discount generalisation

### PR DESCRIPTION
**Pull request by charbeljc**
Refactor module `product_visible_discount` to :
- handle discounts of 100% (actual price and/or base price == 0.0)
- handle cases when base price is computed from another pricelist instead of `product.list_price`

### CODE REVIEW

File: addons/product_visible_discount/product_visible_discount.py [hunk 1]
```diff
@@ -38,40 +39,60 @@ class sale_order_line(osv.osv):
...
             if rule_id:
                 item_base = item_obj.read(cr, uid, [rule_id], ['base'])[0]['base']
                 if item_base > 0:
                     price_type = price_type_obj.browse(cr, uid, item_base)
                     field_name = price_type.field
                     currency_id = price_type.currency_id
+                    product_read = product_obj.read(cr, uid,
+                                                    [product_id], [field_name],
+                                                    context=context)[0]
+                    base_price = product_read[field_name]
```
Comment (mdietrichc2c) [inline addons/product_visible_discount/product_visible_discount.py]: It's from the old code as well, but is there a reason not to use the previously-created `product` browse_record to retrieve the field?

File: addons/product_visible_discount/product_visible_discount.py [hunk 2]
```diff
@@ -38,40 +39,60 @@ class sale_order_line(osv.osv):
...
+                else:
+                    assert item_base == -1
+                    # based on other pricelist
+                    base_price_list_id = item_obj.read(cr, uid,
+                                                       [rule_id],
+                                                       ['base_pricelist_id'])[0]['base_pricelist_id'][0]
```
Comment (mdietrichc2c) [inline addons/product_visible_discount/product_visible_discount.py]: Maybe regroup this read with the one at the top made on `'base'` ?

File: addons/product_visible_discount/product_visible_discount.py [hunk 3]
```diff
@@ -38,40 +39,60 @@ class sale_order_line(osv.osv):
...
+                    pl_obj = self.pool.get('product.pricelist')
```
Comment (mdietrichc2c) [inline addons/product_visible_discount/product_visible_discount.py]: The `self.pool.get()` would be better at the top of the method, next to all the others.

File: addons/product_visible_discount/product_visible_discount.py (hunk 1 above)
Comment (charbeljc) [inline addons/product_visible_discount/product_visible_discount.py]: @mdietrichc2c no reason apart patch minimization ... But yes, this is ugly

File: addons/product_visible_discount/product_visible_discount.py [hunk 4]
```diff
@@ -38,40 +39,60 @@ class sale_order_line(osv.osv):
...
+                    base_pl = pl_obj.browse(cr, uid, [base_price_list_id])[0]
```
Comment (mdietrichc2c) [inline addons/product_visible_discount/product_visible_discount.py]: `base_pl = pl_obj.browse(cr, uid, base_price_list_id, context=context)` would be better.

File: addons/product_visible_discount/product_visible_discount.py [hunk 5]
```diff
@@ -80,29 +101,36 @@ def get_real_price(res_dict, product_id, qty, uom, pricelist):
         result=res['value']
         pricelist_obj=self.pool.get('product.pricelist')
         product_obj = self.pool.get('product.product')
+
         if product and pricelist and self.pool.get('res.users').has_group(cr, uid, 'sale.group_discount_per_so_line'):
-            if result.get('price_unit',False):
+            if result.get('price_unit',False) is not False:
```
Comment (mdietrichc2c) [inline addons/product_visible_discount/product_visible_discount.py]: Why was this modified?

File: addons/product_visible_discount/product_visible_discount.py [hunk 6]
```diff
@@ -80,29 +101,36 @@ def get_real_price(res_dict, product_id, qty, uom, pricelist):
...
             list_price = pricelist_obj.price_rule_get(cr, uid, [pricelist],
                     product.id, qty or 1.0, partner_id, context=pricelist_context)
 
+            new_list_price, currency_id = get_real_price_curency(list_price, product.id, qty, uom, pricelist)
+
             so_pricelist = pricelist_obj.browse(cr, uid, pricelist, context=context)
 
-            new_list_price, currency_id = get_real_price_curency(list_price, product.id, qty, uom, pricelist)
-            if so_pricelist.visible_discount and list_price[pricelist][0] != 0 and new_list_price != 0:
+            if so_pricelist.visible_discount: # and list_price[pricelist][0] != 0 and new_list_price != 0:
```
Comment (mdietrichc2c) [inline addons/product_visible_discount/product_visible_discount.py]: Comment should be removed.

File: addons/product_visible_discount/product_visible_discount.py [hunk 7]
```diff
@@ -80,29 +101,36 @@ def get_real_price(res_dict, product_id, qty, uom, pricelist):
...
                     ctx = context.copy()
                     ctx['date'] = date_order
                     new_list_price = self.pool['res.currency'].compute(cr, uid,
-                        currency_id.id, so_pricelist.currency_id.id,
-                        new_list_price, context=ctx)
-                discount = (new_list_price - price) / new_list_price * 100
+                                                                       currency_id.id,
+                                                                       so_pricelist.currency_id.id,
+                                                                       new_list_price, context=ctx)
+
```
Comment (mdietrichc2c) [inline addons/product_visible_discount/product_visible_discount.py]: Not PEP-8 compliant; you can do a line return before `cr`.

File: addons/product_visible_discount/product_visible_discount.py [hunk 8]
```diff
@@ -80,29 +101,36 @@ def get_real_price(res_dict, product_id, qty, uom, pricelist):
...
+                discount = (new_list_price - price) / new_list_price * 100 \
+                    if new_list_price else 100.0 if price else 0.0
```
Comment (mdietrichc2c) [inline addons/product_visible_discount/product_visible_discount.py]: Some parentheses or putting the `if price` on top would be a bit more readable, IMHO.

**Comment by gurneyalex** [comment]
automated test needed, as well as the reference to PR on upstream.

File: addons/product_visible_discount/product_visible_discount.py (hunk 1 above)
Comment (charbeljc) [inline addons/product_visible_discount/product_visible_discount.py]: ok, cleaning this up ...

File: addons/product_visible_discount/product_visible_discount.py (hunk 5 above)
Comment (charbeljc) [inline addons/product_visible_discount/product_visible_discount.py]: because we had to deal with gratis products and needed to discriminate cases where price_unit is unset (False) vs set to 0.0

**Comment by lepistone** [comment]
This appears to conflict with recent changes upstream in the same module: https://github.com/odoo/odoo/commit/bb90e55fb5f8e38f3c80495a96e747306de9edd4 and https://github.com/odoo/odoo/commit/661f76c94a97ac6304822edeae270d538efaf7e6.

I will investigate further, let's stand by here for now.

**Comment by pedrobaeza** [comment]
No answer, so I close this.
